### ISSUE TITLE
Link Switcher Behaviour in SC 8.2

### ISSUE THREAD

**Issue by mikeedwards83** [sergeyshushlyapin/Sitecore.FakeDb#153]
Hi 

I am noticing different behaviour for the Link Switcher in 8.2 when compared to 8.1.

If you take a look at the following tests:
![image](https://cloud.githubusercontent.com/assets/123856/20869000/1299d4b2-ba61-11e6-9d55-e4c6f638c6b2.png)

http://build.glass.lu/viewLog.html?buildId=2640&buildTypeId=bt25&tab=testsInfo

You can see that the first pair run against 8.2 have failed  but the second pair run against 8.1 pass.

The code for the tests are here:
https://github.com/mikeedwards83/Glass.Mapper/blob/8f4506a14b7c1cd3f9179ed6459741c73fa7ea1a/Tests/Integration%20Tests/Sitecore/Glass.Mapper.Sc.Integration/DataMappers/SitecoreFieldLinkMapperFixture.cs#L163

https://github.com/mikeedwards83/Glass.Mapper/blob/8f4506a14b7c1cd3f9179ed6459741c73fa7ea1a/Tests/Integration%20Tests/Sitecore/Glass.Mapper.Sc.Integration/DataMappers/SitecoreFieldLinkMapperFixture.cs#L231

Is this a known issue or is the test wrong?

Thanks

Mike

**Comment by pveller** [comment]
Hey @mikeedwards83, what assertion(s) fail against 8.2?

**Comment by sergeyshushlyapin** [comment]
@mikeedwards83 you're right, this part stopped working and unfortunately (or likely?) there will be more broken things with new Sitecore versions. The old _static manager / abstract provider_ paradigm is being deprecated and replaced with the brand new [Sitecore DI](http://kamsar.net/index.php/2016/08/Dependency-Injection-in-Sitecore-8-2/).

That is how the `LinkManager.GetItemUrl()` method looks like now (SC8.2+):
```csharp
public static string GetItemUrl(Item item, UrlOptions options)
{
    return LinkManager.Instance.Value.GetItemUrl(item, options);
}
```

where `Instance` is of type `Sitecore.Abstractions.BaseLinkManager, Sitecore.Kernel` which can be mocked and injected via DI container. There is no *Provider* used anymore so that FakeDb's [SwitchingLinkProvider](https://github.com/sergeyshushlyapin/Sitecore.FakeDb/blob/87d2206e07dad5ee8e6e557f3e687ec8402d1b40/src/Sitecore.FakeDb/Links/SwitchingLinkProvider.cs) is never called.

**Comment by sergeyshushlyapin** [comment]
The expectation from Sitecore is to stop relying on the static managers and start injecting new abstractions into classes via constructor. That sounds right but might be pretty tricky to achieve, at least for early 8.2 versions where there are still numerous 'old' references to statics classes:

![image](https://cloud.githubusercontent.com/assets/4719165/20880238/c1d65f94-badf-11e6-94f0-56cdc639af82.png)

I did a [small POC](https://github.com/sergeyshushlyapin/Sitecore.FakeDb/commit/ba5975f80cae70f1f0ee316bb1a81f13be587116), nothing prevents FakeDb from registering own *Switching-* managers using new DI. That allows to continue using the well-known Sitecore `Switcher<T>` to mock the provider (or manager) instance:

```csharp
BaseLinkManager current = Substitute.For<BaseLinkManager>();
current.GetItemUrl(item).Returns(expected);
using (new Switcher<BaseLinkManager>(current))
{
    sut.GetItemUrl(item)
        .Should().Be(expected);
}
```

The unpleasant thing is that FakeDb have to be upgraded to .NET 4.5.2 throwing away compatibility with all the version prior to 8.2.

**Comment by mikeedwards83** [comment]
You might have to go to the solution Glass uses which is a different
compile per SC version.

It then becomes trickier to manage but I am happy to explain how we do it.

**Comment by sergeyshushlyapin** [comment]
@mikeedwards83, I would appreciate if you could briefly describe the idea and share some links so that I could research it deeply. Thanks.

**Comment by mikeedwards83** [comment]
The hardest bit is setting up the project files for compile. I have a single Proj file that gets include by each project in Glass that controls things live .net version number and Sitecore DLL locations:

https://github.com/mikeedwards83/Glass.Mapper/blob/master/Configuration.proj

This is included here:

https://github.com/mikeedwards83/Glass.Mapper/blob/master/Source/Glass.Mapper.Sc/Glass.Mapper.Sc.csproj#L45

And you can see the references to SC assemblies takes in a variable:
https://github.com/mikeedwards83/Glass.Mapper/blob/master/Source/Glass.Mapper.Sc/Glass.Mapper.Sc.csproj#L59

We also use preprocessor directives to fix problems in each version, these are defined in the Configuraiton.proj file:

https://github.com/mikeedwards83/Glass.Mapper/blob/master/Source/Glass.Mapper.Sc/GlassHtml.cs#L879

Finally all the different assemblies for reach SC version get compiled into a single Nuget. We then use a powershell script when  Glass is installed to determine which version to reference. The downside is that the PS script is not supported in Nuget V3 (we have not solved this problem yet):

https://github.com/mikeedwards83/Glass.Mapper/blob/master/Powershell/Glass.Mapper.Sc/Install.ps1

**Comment by pveller** [comment]
I now remembered :) that I have set up a very complex build for SCORE and we were releasing and continuously testing different nugets for different Sitecore versions as well. Mostly a combination of msbuild variables and powershell scripts. It was up to the user though to pull down the right package from the feed. We used a versioning scheme with Sitecore versions in it. Like SCORE.1.2-8.0.xxxx where xxxx is the internal update version (timestamp)
